### DISCUSSION TITLE
Can I suppress `run_always` from `apply --verbose`?

### DISCUSSION THREAD

**Discussion by felipecrs** [twpayne/chezmoi#1666] [Q&A] [answered]
Because, for me, it does not make sense. I like to always run with `--verbose`, but as it's a `run_always` script I would prefer to omit its diff.

**Comment by twpayne** [comment, marked as answer]
Currently it's only possible to exclude scripts entirely from the diff with:

```toml
[diff]
    exclude = ["scripts"]
```

in your config. It's not currently possible to differentiate between `run_` and `run_once_` scripts.

**Reply by felipecrs** [reply]
Excluding scripts entirely kinda help. I will apply this to my dotfiles. Thank you!

**Reply by felipecrs** [reply]
@twpayne this does not seem to work for the `apply` command, only for the `diff` command:

```console 
❯ cat ~/.config/chezmoi/chezmoi.yaml
diff:
  exclude:
    - scripts

❯ chezmoi diff

❯ chezmoi apply --verbose --dry-run
diff --git a/always_test.sh b/always_test.sh
index e69de29bb2d1d6434b8b29ae775ad8c2e48c5391..13fdcce1e520923baf8faa0371c5f0bba29b66b2 100755
--- a/always_test.sh
+++ b/always_test.sh
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+echo test
```

**Reply by twpayne** [reply]
Ah, good point. #1676 should mostly fix this. It applies to all scripts though. It is not currently possible to distinguish between `run_` and `run_once_` scripts.

**Reply by felipecrs** [reply]
That's ok. Usually, in my scripts, I'm very verbose about what they are currently doing, so I believe it's fine. Thank you!

**Comment by Lockszmith-GH** [comment]
Was looking for a solution for suppressing 'always run' scripts, and still displaying changes in `run_once` scripts, and found this thread.

@twpayne - do you have plans on implementing a filter that will show script modifications and not 'planned runs'?

For now, I do a mix of `chezmoi status --exclude scripts` and a `chezmoi git -- ls-files _home.base/.chezmoiscripts/ --modified` (`_home.base` is where my root dir is)

**Reply by twpayne** [reply]
> @twpayne - do you have plans on implementing a filter that will show script modifications and not 'planned runs'?

I didn't have plans but since @felipecrs and you requested the feature (which is a very reasonable request as `chezmoi diff` output should be interesting to the user and not include any unnecessary fluff), I got [nerd-sniped](https://xkcd.com/356/) and am now working on this in the https://github.com/twpayne/chezmoi/compare/exclude-always branch. This is very much a work in progress with many failing tests as it's lead to an overhaul of chezmoi's `--include` and `--exclude` options.

**Reply by Lockszmith-GH** [reply]
Awesome! I'll follow closely, thank you so much for taking an interest in this request.

(Beware of trucks, and... does a software developer scores more or less than a mathematician?)

EDIT: WOW that's a major refactoring, crossing my fingers that doesn't turn into a deadly truck indeed. Sorry for repeating myself, but... Thank you so much.

**Reply by Lockszmith-GH** [reply]
This is now a permanent fixture in my config. Thanks!
```
[diff]
    exclude = ["always"]

[status]
    exclude = ["always"]
```

**Reply by felipecrs** [reply]
For some reason this works for `chezmoi diff` and `chezmoi status`, but it doesn't work for `chezmoi apply --verbose` (i.e. the `always` scripts are still shown in the apply diff).

**Reply by twpayne** [reply]
Great catch @felipecrs! Fix in #2565.
